### PR TITLE
297 Pagination based on paging_state as last page last transaction address GraphQl API

### LINKED ISSUE #297: Use pagination state in GraphQL API
### Problem to solve

GraphQL API `transactionChains` uses pagination mechanism using number.
However it may create confusion which number to use 0 or 1 for the first page causing different results due to implementation `Enum.at/2`.
Also we would like to avoid memory consumption by loading the entire chain.

Related to #296

### Solution

We can leverage the `pagination_state` returned from the DB and provided by the `GetTransactionChain` message.

The parameter will be then be the last address retrieved as the starting point on the chain to be retrieved.

However, for `networkTransaction` API we can still use the page as a number greater than 0, as the state will be different for several chains.

### Development Tasks

- [x] Breaking existing code in to definitions for re-usability for newer approach
- [x] Creating a New Pagination approach based on `paging_address` instead page number where `paging_address` is page's last transaction address
- [x] Add new function resolver to fetch transaction chain based on `paging_address` instead of loading all transactions and slicing down to pages
- [x] Add `networkTransactions` API `page` attribute  validation 
- [x] Add or Update Relevant Unit Test cases for Transaction Chain

### PR DESCRIPTION
**Pull request by blackode**
# Description

Before this change, to get the specific page level transaction_chain, we are loading whole transactions and slicing them by the page size and retrieving the relevant page transactions. 

In this change, the `page` attribute which was `number` now converted to `pagingAddress` of type `string` is now the address of the last transaction in the page which act as a `paginate_state` for next page.

Here, we are loading specific `paging_state` based page size transactions only. 

Fixes #297 #296 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?
## Steps
1. Select the address whose chain length is greater 10 for an example consider chain_length of 15
2. Now run query `transactionChain(address: "address_selected"){address}`, which gives 10 transactions
3. Make sure it is listed only 10 transactions
4. Pick the last transaction address from above 10 transactions
5. Run the query `transactionChain(address: "address_selected", pagingAddress: "last_transaction_address"){address}`
6. Now above query should give only 5 transactions
7. Now take out the last transaction address from above 5 transactions and repeat the step 5
8. It should give empty list of transactions.

- Test with out paging state based transactions
- Test paging_state based transactions

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic_web/graphql_schema.ex [hunk 1]
```diff
@@ -58,11 +58,19 @@ defmodule ArchethicWeb.GraphQLSchema do
     """
     field :transaction_chain, list_of(:transaction) do
       arg(:address, non_null(:address))
-      arg(:page, :integer)
+      arg(:paging_address, :string)
```
Comment (samuel-uniris) [inline lib/archethic_web/graphql_schema.ex:61]: We need to valid as well if it's a valid address, so the type should be `address`

File: lib/archethic_web/graphql_schema.ex [hunk 2]
```diff
@@ -58,11 +58,19 @@ defmodule ArchethicWeb.GraphQLSchema do
...
 
       resolve(fn args = %{address: address}, _ ->
-        page = Map.get(args, :page, 1)
-        Resolver.paginate_chain(address, page)
+        paging_address = Map.get(args, :paging_address)
+
+        paging_address =
+          if paging_address do
+            Base.decode16!(paging_address)
```
Comment (samuel-uniris) [inline lib/archethic_web/graphql_schema.ex:68]: By validating the paging_address by the type `address` this cast will be not necessary

File: lib/archethic.ex [hunk 3]
```diff
@@ -240,11 +240,34 @@ defmodule Archethic do
   """
   @spec get_transaction_chain(binary()) :: {:ok, list(Transaction.t())} | {:error, :network_issue}
   def get_transaction_chain(address) when is_binary(address) do
+    local_available_nodes = locally_available_nodes(address)
+    get_transaction_chain(local_available_nodes, address)
+  end
+
+  def get_transaction_chain_by_paging_address(address, paging_address) when is_binary(address) do
```
Comment (samuel-uniris) [inline lib/archethic.ex:247]: Could you add typespec and doc as well ?

File: lib/archethic_web/graphql_schema/resolver.ex [hunk 4]
```diff
@@ -112,7 +112,7 @@ defmodule ArchethicWeb.GraphQLSchema.Resolver do
     )
   end
 
-  def network_transactions(type, page) do
+  def network_transactions(type, page) when page >= 1 do
```
Comment (samuel-uniris) [inline lib/archethic_web/graphql_schema/resolver.ex:115]: I think you should also return an error when the page is < 1

**Review by samuel-uniris — CHANGES_REQUESTED** [review]
Thanks for the PR.
Overall is good, just some changes to perform

File: lib/archethic.ex (hunk 3 above)
Comment (blackode) [reply lib/archethic.ex]: Updated the function with doc and typespe

File: lib/archethic_web/graphql_schema/resolver.ex (hunk 4 above)
Comment (blackode) [reply lib/archethic_web/graphql_schema/resolver.ex]: Introduced new `page_type`

File: lib/archethic.ex [hunk 5]
```diff
@@ -295,6 +269,38 @@ defmodule Archethic do
...
+           paging_state: Keyword.get(options, :paging_state)
+         }) do
+      {:ok, %TransactionList{transactions: transactions}} ->
+        {:ok, transactions}
+
+      {:error, _} ->
+        transaction_chain_by_paging_address(rest, address, options)
+    end
+  end
+
```
Comment (samuel-uniris) [inline lib/archethic.ex:303]: I think you are missing the case when there is not nodes available.
This is causing the dialyzer issue I think

Comment (blackode) [reply lib/archethic.ex:303]: Yeah, I missed the case when nodes list is empty `[]` 
Thanks :)

**Comment by samuel-uniris** [comment]
I think it could be great also to apply the `page` type in `transactions` queries (locally stored)

File: lib/archethic_web/graphql_schema.ex [hunk 6]
```diff
@@ -105,7 +107,7 @@ defmodule ArchethicWeb.GraphQLSchema do
     """
     field :network_transactions, list_of(:transaction) do
       arg(:type, :transaction_type)
```
Comment (samuel-uniris) [inline lib/archethic_web/graphql_schema.ex:109]: I could be great also to add field requirement on the type

**Review by samuel-uniris — APPROVED** [review]
It's working thanks 👍

**Review by samuel-uniris — APPROVED** [review]

**Pull request merged**